File: 2428-equal-row-and-column-pairs/solution.cpp
class Solution {
public:
    int equalPairs(vector<vector<int>>& grid) {
      int n = grid.size();
      int count = 0;
      map<vector<int>, int> row;

      for (auto &it: grid)
        ++row[it];

      for (int i = 0; i < n; ++i) {
        vector<int> col;

        for (int j = 0; j < n; ++j)
          col.emplace_back(grid[j][i]);

        count += row[col];
      }
      return count;
    }
};

File: 2428-equal-row-and-column-pairs/solution.ts
const equalPairs = (grid: number[][]): number => {
  const n = grid.length;
  let count = 0;
  const rowFrequency = new Map<string, number>();

  for (let i = 0; i < n; ++i) {
    const row = grid[i].join('-');
    rowFrequency.set(row, (rowFrequency.get(row) || 0) + 1);
  }

  for (let i = 0; i < n; ++i) {
    const column: number[] = [];

    for (let j = 0; j < n; ++j) {
      column.push(grid[j][i]);
    }
    const columnString = column.join('-');

    if (rowFrequency.has(columnString)) {
      count += rowFrequency.get(columnString);
    }
  }

  return count;
};
